documentation/specs/rar-as-service.md

```diff
@@ -0,0 +1,217 @@
+
+If the connection is successful, we can use this connection for execution of RAR task. The node is guaranteed to have all required properties since they must be encoded in name of the RAR node.
+
+### Start new node
+
+This step will create new process which will act as RAR node. It will also pass necessary information to the node to know what its settings are (reusable, ...). Node will be another instance of the MSBuild.exe which will have set parameter **nodeMode** to some specific value. Probably it will be text value, so we can easily recognize which node is RAR node and which is normal MSBuild node.
+
+There will be mechanism which will ensure that we don't create two RAR nodes at once. It will be done via Mutex (as it is done in [Roslyn](https://github.com/dotnet/roslyn/blob/838874b1b817db84ce146bef690cc95a39c213a5/src/Compilers/Server/VBCSCompiler/BuildServerController.cs#L143)). The name of the mutex has to encode in some way, all the necessary information needed to decide if it is only one RAR node for user. There should be at least: User name, If is administrator and maybe some initial settings for the node. An example of that name could be: `MSBuild.RAR.ostorc.7`, where **MSBuild.RAR** is prefix, **ostorc** is username under which the MSBuild was invoked and **7** represents encoded settings (flagged enum).
+
+### Execute RAR task
+
+Execution of RAR task should be done in same way as it is done now.
+
+There is already some layer of separation between Task interface and actual execution method. We will leverage this, and put the decision logic if to run locally or not into the "wrapper" method and so we will not have to modify this and in server-side execution we will directly call the internal execution method.
+
+There is one big concern and that is how to handle multiple requests at once. As right now, RAR task is not prepared for multi-thread use.
+
+### Shutdown RAR node
+
+If the user want's the node to not be reused, we have the ensure that node will be killed after the build ends. This should be done after the compilation end by main MSBuild node.
+
+The RAR node, also has to support accepting of already established commands for MSBuild nodes (for example Shutdown command). This will be done by creating two pipes inside node, one will be for communication 
+
+### Execute task in MSBuild node
+
+User opted out of using the RAR nodes so we will execute the RAR task in the MSBuild node (as it is right now).
+
+### Other
+
+The new RAR node will not count toward total maximum CPU count specified by _/maxCpuCount_ switch, since the RAR task is taxing on IO operations not so much on CPU time. If we took one node from each instance of MSBuild it would lead to drastic decrease in performance.
+
+The RAR task will be affected by  _/m_ switch. When we run in single node mode, it will implicitly say that we want to run task inside current process. User would have to explicitly say that he wants to run with RAR node.
```

Comment:
```suggestion
The RAR task will be affected by the _/m_ switch. When we run in single node mode, it will implicitly say that we want to run the task inside current process. User would have to explicitly say that they want to use the RAR node.
```

And a question: Assuming that the feature is going to be opt-in, even after it's fully implemented, isn't the above statement true regardless of _/m_? I.e. if not opted in, it's going to behave just like today, running the task in the process building the respective project, and if opted in it will use the RAR node always?

Reply:
Text updated. That paragraph is valid if the feature is opt-out. Added note, that the default state of this feature is not yet decided.